engine/consensus/dkg/dkg_controller_test.go

```diff
@@ -0,0 +1,255 @@
+	}
+
+	// run each test case
+	for _, tc := range testCases {
+		// gn is the minimum number of good nodes required for the DKG protocol
+		// to go well
+		gn := tc.totalNodes - optimalThreshold(tc.totalNodes)
+
+		t.Run(fmt.Sprintf("%d/%d nodes", gn, tc.totalNodes), func(t *testing.T) {
+			testDKG(t, tc.totalNodes, gn, tc.phaseDuration)
+		})
+	}
+}
+
+// ATTENTION: This test seems to indicate that nodes end up with a different set
+// of public keys when the phase duration is short.
+func TestDKGShortDuration(t *testing.T) {
+	testDKG(t, 20, 20, 2*time.Second)
+}
+
+func testDKG(t *testing.T, totalNodes int, goodNodes int, phaseDuration time.Duration) {
+	nodes := initNodes(t, totalNodes, phaseDuration)
+	gnodes := nodes[:goodNodes]
+
+	// Start all the good nodes in parallel
+	for _, n := range gnodes {
+		go func(node *node) {
+			err := node.run()
+			require.NoError(t, err)
+		}(n)
+	}
+
+	// Wait until they are all shutdown
+	wait(t, gnodes, 5*phaseDuration)
+
+	// Check that all nodes have agreed on the same set of public keys
+	checkArtifacts(t, gnodes, totalNodes)
+}
+
+// Initialise nodes and communication channels.
+func initNodes(t *testing.T, n int, phaseDuration time.Duration) []*node {
+	// Create the channels through which the nodes will communicate
+	channels := make([]chan DKGMessage, 0, n)
+	for i := 0; i < n; i++ {
+		channels = append(channels, make(chan DKGMessage, 5*n*n))
+	}
+
+	nodes := make([]*node, 0, n)
+
+	// Setup
+	for i := 0; i < n; i++ {
+		seed := make([]byte, 20)
+		_, _ = rand.Read(seed)
+
+		processor := &processor{
+			id:       i,
+			channels: channels,
+		}
+
+		dkg, err := crypto.NewJointFeldman(n, optimalThreshold(n), i, processor)
+		require.NoError(t, err)
+
+		controller := NewController(
+			dkg,
+			seed,
+			channels[i],
+			zerolog.New(os.Stderr).With().Int("id", i).Logger())
+
+		node := newNode(i, controller, phaseDuration)
+
+		nodes = append(nodes, node)
+	}
+
+	return nodes
+}
+
+// Wait for all the nodes to reach the SHUTDOWN state, or timeout.
+func wait(t *testing.T, nodes []*node, timeout time.Duration) {
+
+	timer := time.After(timeout)
+
+	for {
+		select {
+		case <-timer:
+			t.Fatal("TIMEOUT")
+		default:
+			done := true
+			for _, node := range nodes {
+				if node.controller.GetState() != Shutdown {
+					done = false
+					break
+				}
+			}
+			if done {
+				return
+			}
+			time.Sleep(1 * time.Second)
+		}
+	}
+}
+
+// Check that all nodes have produced the same set of public keys
+func checkArtifacts(t *testing.T, nodes []*node, totalNodes int) {
+	for i := 1; i < len(nodes); i++ {
+		require.NotEmpty(t, nodes[i].controller.privateShare)
+		require.NotEmpty(t, nodes[i].controller.publicShare)
+		require.Len(t, nodes[i].controller.publicKeys, totalNodes)
+
+		for j := 0; j < totalNodes; j++ {
+			if !reflect.DeepEqual(nodes[0].controller.publicKeys[j], nodes[i].controller.publicKeys[j]) {
+				t.Fatalf("pubs[%d] differ: %s, %s", j,
```

Comment:
I think logging `i` helps.
```suggestion
				t.Fatalf("node %d has a different pubs[%d] than node 0: %s, %s", i, j,
```